From 37840255aefab349a51198b9648e50483755c80a Mon Sep 17 00:00:00 2001
From: Jasper Orschulko <Jasper.Orschulko@iris-sensing.com>
Date: Thu, 29 Apr 2021 16:38:09 +0200
Subject: [PATCH] filter denied virtual interfaces when adding addresses

netlink does not send an interface change notification when setting up a virtual
interface such as eth0:0: that information is only available in RTM_NEWADDR
messages.

It is therefore needed to check if a new address does belong to an allowed
interface label before actually adding it to the hardware interface.

Upstream-Status: Pending: http://jira.irisgmbh.local/browse/APC-3065
---
 avahi-core/iface-linux.c | 23 +++++++++++++++++++++--
 avahi-core/iface.c       | 27 ++++++++++++++++-----------
 avahi-core/iface.h       |  1 +
 3 files changed, 38 insertions(+), 13 deletions(-)

diff --git a/avahi-core/iface-linux.c b/avahi-core/iface-linux.c
index e116c7b..775e2ee 100644
--- a/avahi-core/iface-linux.c
+++ b/avahi-core/iface-linux.c
@@ -183,6 +183,7 @@ static void netlink_callback(AvahiNetlink *nl, struct nlmsghdr *n, void* userdat
         size_t l;
         AvahiAddress raddr, rlocal, *r;
         int raddr_valid = 0, rlocal_valid = 0;
+        char label[IFNAMSIZ + 1];
 
         /* We are only interested in IPv4 and IPv6 */
         if (ifaddrmsg->ifa_family != AF_INET && ifaddrmsg->ifa_family != AF_INET6)
@@ -194,6 +195,8 @@ static void netlink_callback(AvahiNetlink *nl, struct nlmsghdr *n, void* userdat
         if (!(i = avahi_interface_monitor_get_interface(m, (AvahiIfIndex) ifaddrmsg->ifa_index, avahi_af_to_proto(ifaddrmsg->ifa_family))))
             return;
 
+        memset(label, 0, IFNAMSIZ + 1);
+
         /* Fill in address family for our new address */
         rlocal.proto = raddr.proto = avahi_af_to_proto(ifaddrmsg->ifa_family);
 
@@ -229,16 +232,32 @@ static void netlink_callback(AvahiNetlink *nl, struct nlmsghdr *n, void* userdat
 
                     memcpy(raddr.data.data, RTA_DATA(a), RTA_PAYLOAD(a));
                     raddr_valid = 1;
-
                     break;
 
+                case IFA_LABEL: {
+                    size_t label_len;
+
+                    if (RTA_PAYLOAD(a) > IFNAMSIZ) {
+                        label_len = IFNAMSIZ;
+                        avahi_log_warn("interface label longer than IFNAMSIZ !");
+                    }
+                    else
+                        label_len = RTA_PAYLOAD(a);
+                    memcpy(label, RTA_DATA(a), label_len);
+                    label[label_len] = '\0';
+                }
+                break;
+
                 default:
                     ;
             }
-
             a = RTA_NEXT(a, l);
         }
 
+        /* Ignore if label denied in avahi-daemon.conf. */
+        if (strlen(label) != 0 && !avahi_interface_label_is_relevant(i, label))
+            return;
+
         /* If there was no adress attached to this message, let's quit. */
         if (rlocal_valid)
             r = &rlocal;
diff --git a/avahi-core/iface.c b/avahi-core/iface.c
index a68a4b9..ae78387 100644
--- a/avahi-core/iface.c
+++ b/avahi-core/iface.c
@@ -678,25 +678,30 @@ static int avahi_interface_is_relevant_internal(AvahiInterface *i) {
     return 0;
 }
 
-int avahi_interface_is_relevant(AvahiInterface *i) {
+int avahi_interface_label_is_relevant(AvahiInterface *i, const char *label) {
     AvahiStringList *l;
     assert(i);
+    assert(label);
 
     for (l = i->monitor->server->config.deny_interfaces; l; l = l->next)
-        if (strcasecmp((char*) l->text, i->hardware->name) == 0)
+        if (strcasecmp((char*) l->text, label) == 0)
             return 0;
+    if (!i->monitor->server->config.allow_interfaces)
+        return 1;
+    for (l = i->monitor->server->config.allow_interfaces; l; l = l->next)
+        if (strcasecmp((char*)l->text, label) == 0)
+            return 1;
+    return 0;
+}
 
-    if (i->monitor->server->config.allow_interfaces) {
-
-        for (l = i->monitor->server->config.allow_interfaces; l; l = l->next)
-            if (strcasecmp((char*) l->text, i->hardware->name) == 0)
-                goto good;
+int avahi_interface_is_relevant(AvahiInterface *i) {
+    AvahiStringList *l;
+    assert(i);
 
+    if (avahi_interface_label_is_relevant(i, i->hardware->name))
+        return avahi_interface_is_relevant_internal(i);
+    else
         return 0;
-    }
-
-good:
-    return avahi_interface_is_relevant_internal(i);
 }
 
 int avahi_interface_address_is_relevant(AvahiInterfaceAddress *a) {
diff --git a/avahi-core/iface.h b/avahi-core/iface.h
index c3f24af..49f88cf 100644
--- a/avahi-core/iface.h
+++ b/avahi-core/iface.h
@@ -164,6 +164,7 @@ void avahi_interface_free(AvahiInterface *i, int send_goodbye);
 
 void avahi_interface_update_rrs(AvahiInterface *i, int remove_rrs);
 void avahi_interface_check_relevant(AvahiInterface *i);
+int avahi_interface_label_is_relevant(AvahiInterface *i, const char *label);
 int avahi_interface_is_relevant(AvahiInterface *i);
 
 void avahi_interface_send_packet(AvahiInterface *i, AvahiDnsPacket *p);
-- 
2.31.1

